Split up gadgetbridge apps: gbnotify

detached opened the pull request:
This is the first of three apps for splitting up the gbridge widget: #162
The two other apps (call and music control) are on this branch https://github.com/detached/BangleApps/tree/new-gbridge-apps. I want to keep the pull request small, so I will contribute one after the other.

The gbnotify just handles notifications. The widget shows notifications in the same way as the gbridge app, just slightly redesigned.

![20200503_160026](https://user-images.githubusercontent.com/5504749/80916412-c29e1580-8d58-11ea-9a1e-86dfe31992e2.jpg)

The app allows swiping through the last five messages.

![20200503_160100](https://user-images.githubusercontent.com/5504749/80916481-41934e00-8d59-11ea-8753-4d72d875f2fe.jpg)

![20200503_160110](https://user-images.githubusercontent.com/5504749/80916483-435d1180-8d59-11ea-8c68-fefa361e51b8.jpg)

gfwilliams commented on apps/gbnotify/widget.js:
```diff
@@ -0,0 +1,89 @@
+    if(_GB)setTimeout(_GB,0,event);
+  };
+
+  WIDGETS["mywidget"] = {
```
Please could you change `mywidget` to `gbnotify` here? Otherwise other apps that forget too will just overwrite this

detached replied:
Thanks for spotting this!

gfwilliams commented:
Nice - looks good. I like the idea of storing multiple notifications.

detached commented:
What do you think about register an event listener, so other widgets can show notifications with `Bangle.emit()`?

gfwilliams commented:
> What do you think about register an event listener,

I think probably not - realistically if we want a global way to send events we should probably build it into Bangle.js itself to try and get the memory usage down (and also to ensure that it's available for all apps).

detached commented:
Can I do something to get this merged?

gfwilliams commented:
Sorry for the delay - I've been a bit swamped with other stuff.

The thing that's made me think this is nontrivial is that this looks like it doesn't play well with the existing Gadgetbridge code at all, but then it doesn't seem to do the connection icon or battery level reporting. Having something 'modular' that doesn't work with the normal Gadgetbridge code seems like it's just going to really confuse the majority of users.

So I was trying to think of a way of handling this nicely, especially with several users complaining about being short on RAM - we really want to get things as efficient as possible.

It seems to me that there's a bit of duplication in https://github.com/detached/BangleApps/tree/new-gbridge-apps with how notifications are displayed/dismissed... 

I think in the long term I'll make a 'Notifications' library that can be used for displaying notifications. Maybe someone else makes different ones later (like fullscreen notifications, etc) - we could all call the notifications library, which should hopefully save some space and would also allow some customisation of look and feel later on.

But I guess for now we've got two options:

* Completely remove all notifications from the existing gadgetbridge widget and make it handle Battery level reporting and connection state only, and merge in all your widgets
* Document your widgets very clearly so folks know they don't work alongside the `gbridge` widget.

I'd be interested if others can weigh in here about what you think is best?

I like the modular idea and honestly I think the first option is better. I'll do the notification library now anyway as that should be pretty easy.

One minor point in your code.. Any chance you could remove the whole ` WIDGETS["gbnotify"] = {` completely so it doesn't just create a blank area on the widget bar? Same for `gbcall` - and you can actually change the width of your widget so the music widget could hide itself when not playing music and then pop up?

gfwilliams commented:
Ok, `notify` lib (and ability to specify dependencies) just gone in: https://github.com/espruino/BangleApps/commit/519262793d7ef91dfa91612ef72663e918ea1fe9

I nicked your different notification style - hope that's ok?

So that should tidy up the extra widgets quite a bit I think, if you were interested in implementing it?

detached commented:
Hi @gfwilliams,
sry that I didn't respond for that long time, I had quite a ruff time.

> I nicked your different notification style - hope that's ok?
That is great! :)

I just removed the notification rendering code from the widget and use the notify lib instead and also removed the `WIDGETS` code.

gfwilliams commented:
Sorry for the delay replying to this - and hope things are looking better now.

I think with the Messages/Android app this is no longer work keeping open. Gadgetbridge handling is at least Modular now as well.
